### PR TITLE
Workaround python issue41307#30

### LINKED ISSUE #30: UTF-8 encoded mails are not sent
According to the documentation in `smptlib.sendmail`,

        msg may be a string containing characters in the ASCII range, or a byte
        string.  A string is encoded to bytes using the ascii codec, and lone
        \\r and \\n characters are converted to \\r\\n characters.

Since switching to Zope 4, we are unable to send mails that have a UTF-8 encoded body - Even if we transform it to bytes ourselves, `Products.MailHost` converts it back to strings before sending it to `zope.sendmail` and smptlib fails to encode the passed message using ASCII during the second phase of the two phase commit. Our quick fix was to change `zope.sendmail` to encode the message using `utf-8` before passing it to `smtplib.sendmail`.

A really clean way would probably be to inspect the headers (which should be ASCII) and check for a `Content-Type` and `Content-Transfer-Encoding`. Maybe it is even possible to have a multipart message with different encodings (not sure). But since UTF-8 is such a popular choice, we used that for now.

I am not sure if the fix should happen in `zope.sendmail` or somewhere further up like `Products.MailHost`. If it should be `zope.sendmail`, I could try to supply a PR. Otherwise I would need to be pointed to the right direction.

### PR DESCRIPTION
**Pull request by d-maurer**
This PR switches message serialization from `str` to `bytes` for Python 3. It should fix #30 .

The PR employs heavy workarounds for problems/weaknesses that should be solved elsewhere:
 * "https://bugs.python.org/issue41307" to be fixed in Python
 * `zope.sendmail.delivery` insists on messages being serialized as `str` (this does not fit well with Python's `SMTP.send`) - should be fixed in `zope.sendmail`
 * `zope.sendmail.queue` reads messages as text, instead of as bytes (this does not fit well with Python's `SMTP.send`) -- should be fixed in `zope.sendmail`.

@viktordick has meanwhile enhanced `zope.sendmail` (not yet released) such that the `zope.sendmail` workarounds could be removed. Instead, `Products.MailHost` now depends on the enhanced `zope.sendmail` version.

### CODE REVIEW

**Comment by d-maurer** [comment]
What happened to the Travis tests? I expect still some tests to fail for Python 3.5.

**Comment by jugmac00** [comment]
@d-maurer 3.5 and the lint/flake8 envs still fail, but it looks like there is a disruption between the communication between Travis and GitHub.

https://travis-ci.org/github/zopefoundation/Products.MailHost/builds/708773718

**Comment by d-maurer** [comment]
Jürgen Gmach wrote at 2020-7-16 08:20 -0700:
>@d-maurer 3.5 and the lint/flake8 envs still fail

Thank you for the link!

The "lint" env failure is a general flake8 integration problem
and has nothing to do with my changes:
```
File "/home/travis/build/zopefoundation/Products.MailHost/.tox/flake8/lib/python3.6/site-packages/flake8_isort.py", line 3, in <module>

    from isort import SortImports

ImportError: cannot import name 'SortImports'
```
This looks like the well known `isort` version 5 problem.


The 3.5 failure is expected. My workaround for Python issue 41307
uses a feature available for Python 3.6+.
As a consequence, some tests will fail (and I must add
a "skipUnless(Python 3.6+)".

**Comment by jugmac00** [comment]
Yes, the `isort` failures are unrelated.

Currently, `flake8-isort` is broken in combination with `isort 5` as the latter changed its API. 

Also, some options are now deprecated and the algorithm about how to sort or how to classify what is third party imports has changed.

**Comment by dataflake** [comment]
Dieter, just skipping a test on Python 3.5 won't help the fact that the package as a whole is no longer compatible with Python 3.5 if you use Python 3.6-only features.

**Comment by d-maurer** [comment]
Jens Vagelpohl wrote at 2020-7-16 08:34 -0700:
>Dieter, just skipping a test on Python 3.5 won't help the fact that the package as a whole is no longer compatible with Python 3.5 if you use Python 3.6-only features.

I will look which tests break.

If only the one test breaks which I have created to
reproduce @viktordick 's problem, then this should be no problem
-- it does not work anyway with the current `MailHost` - independent
of the Python 3 version. But, I expect that another test will break
also involving an "8bit" encoding. For this one, I expect that
even though the test succeeds with the current `MailHost` it would not
be possible to actually sent the corresponding message - again
giving the problem reportet by Viktor. If this should turn out to be
the case, that test, too, could be skipped for Python 3.5.

**Comment by d-maurer** [comment]
Dieter Maurer wrote at 2020-7-16 22:17 +0200:
>Jens Vagelpohl wrote at 2020-7-16 08:34 -0700:
>>Dieter, just skipping a test on Python 3.5 won't help the fact that the package as a whole is no longer compatible with Python 3.5 if you use Python 3.6-only features.
>
>I will look which tests break.
>
>If only the one test breaks which I have created to
>reproduce @viktordick 's problem, then this should be no problem
>-- it does not work anyway with the current `MailHost` - independent
>of the Python 3 version. But, I expect that another test will break
>also involving an "8bit" encoding. For this one, I expect that
>even though the test succeeds with the current `MailHost` it would not
>be possible to actually sent the corresponding message - again
>giving the problem reportet by Viktor. If this should turn out to be
>the case, that test, too, could be skipped for Python 3.5.

Exactly, the two expected tests fail:

 * `testMailHost.TestMailHostQueuing.test_8bit_special`
   this is the one I have created to reproduce #30

 * `testMailHost.TestMailHosttest.Explicit8bitEncoding`
   is similar to #30 (also involving an 8bit content transfer encoding)

As expected, Python3 cannot send the messages corresponding
to those tests due to the problem described in #30
(even though the latter tests succeeds).

Below the transscript for Python 3.6, `Products.MailHost==4.9`
trying to send the message from the `Explicit8bitEncoding` test:
```
>>> from Products.MailHost.MailHost import MailHost
>>> mh = MailHost()
>>> mt = 'Date: Sun, 27 Aug 2006 17:00:00 +0200\n\nA M\xc3\xa9ssage'
>>> mh.send(mt, mfrom='sender@example.com', mto='dieter@localhost', encode='8bit', immediate=True)
Traceback (most recent call last):
...
  File "/home/dieter/tmp/py3/lib/python3.6/site-packages/zope/sendmail/mailer.py", line 115, in send
    connection.sendmail(fromaddr, toaddrs, message)
  File "/usr/lib/python3.6/smtplib.py", line 855, in sendmail
    msg = _fix_eols(msg).encode('ascii')
UnicodeEncodeError: 'ascii' codec can't encode characters in position 167-168: ordinal not in range(128)
```

Thus, we do not really lose something when those tests are skipped
for Python 3.5.

**Comment by icemac** [comment]
Alternatively we could also drop support for Python 3.5 as its EOL is in September 2020.

**Comment by icemac** [comment]
Zope 5 already dropped support for Python 3.5.

**Comment by dataflake** [comment]
Zope 5 did, but not Zope 4. If we dropped Python 3.5 support from this one package you'd immediately have to drop Zope 4 support as well.

**Comment by icemac** [comment]
Will we keep supporting Python 3.5 on Zope 4 after its EOL and when Zope 5 is finally released?

**Comment by dataflake** [comment]
The roadmap we came up with at https://zope.readthedocs.io/en/latest/roadmap.html says Zope 4 supports Python 2.7 and 3.5, among others. In my opinion we should stick to that.

Keep in mind that the "pain" of supporting Python 3.5 falls away the moment Zope 5 is released because the roadmap already tells people that Zope 4 will move to bugfix-only support when Zope 5 is out.

Since Zope 4 is bugfix-only at that point the barrier to create major releases in other packages - like this one - is a lot lower. Those major releases can then drop support for anything that's not supported by Zope 5.

**Comment by icemac** [comment]
Okay. Agreed.

**Comment by d-maurer** [comment]
> Alternatively we could also drop support for Python 3.5 as its EOL is in September 2020.

This PR works around bugs/weaknesses of (among others) `zope.sendmail`:
Python's `smtplib.SMTP.sendmail` insists that the message body is either a sequence of bytes or a pure ASCII `str`. Parts of `zope.sendmail` try to support message bodies which are either bytes or `str` to be converted into `bytes` via `utf-8` encoding (the latter is likely accidental - a remaining from Python2 unicode support). But support for `bytes` message bodies breaks due to a silly "message-id" determination and support not non ASCII `str` breaks due to the "smtplib" limitations.

I will soon propose a PR for `zope.sendmail` fixing those issues. It will then be able to handle `bytes` message bodies and `str` message bodies where the latter are encoded via `utf-8` before transfer to `sendmail`. Such an encoding happens already now -- during message queueing.

This will fix @viktordick 's problem as well (the latter part is actually his original proposal). It may allow us to leave `Products.Mailhost` untouched - even though from a conceptual viewpoint it is more correct to represent a message body as bytes than as `utf-8` encodable `str`. There will still be a corner case: `email.message.Message` uses unicode surrogates to represent non ASCII bytes in messages created via `message_from_bytes`. Those are not encodable via `utf-8`. As a consequence, `zope.sendmail` will not be able even after the fixes to send such messages - unless they have been formerly converted into bytes (this PR would ensure this in `Products.MailHost`).

**Comment by d-maurer** [comment]
Michael Howitz wrote at 2020-7-17 07:47 -0700:
>Will we keep supporting Python 3.5 on Zope 4 after its EOL and when Zope 5 is finally released?

Let me stress again that the PR will not prevent
Python 3.5 support by `Products.MailHost`.
It will only cause some messages which can currently not be sent
in all Python 3 versions to work in Python 3.6+ and fail earlier
in Python 3.5 - sufficiently early that the failure can be catched
by 2 of our tests (which therefore need to be skipped for Python 3.5).

**Comment by d-maurer** [comment]
Viktor Dick wrote at 2020-7-19 02:40 -0700:
>@viktordick commented on this pull request.
> ...
>However, it seems to me that passing explicit `encoding` and `charset` parameters to `send` still yields some surprising behavior and inconsistencies.

Those are for simple situations - a single message without "multipart".

If you have complex messages do not use them. Do the necessary
operations at the `email.message.Message` level (or even higher up).

>Maybe it is just me not understanding the idea. But I guess we should start with an exhaustive docstring for `send` and `_mungeHeaders` explaining the intended effect of all parameters (but especially these two), including what an input of bytes vs. unicode entails.

I have never objections against better documentation.

>I guess if the input of `send` is bytes, we should not decode the whole message with one `charset` because some parts might have an explicit `charset` defined that differs from the default.

For Python 3 (where "bytes" is an explicit choice), I would say
that a message provided as `bytes` should not be transformed at all:
it already has the logical external format, We may just add/replace
some information (e.g. the target addresses).

> ...
>Alternatively, passing a `charset` value could force a re-encoding to keep the strategy of "passed parameters overwrite already set values in the message"

I do not see real value to recode parts of a message due to an externally
provided `charset` (or `encode`): those are low level implementation
details which should already have been correct when the message
has been constructed.

This contrasts with e.g. "To": you may want to send the same
message to different recipients; thus, it is very convenient
to able to control the "To:" header from the outside.

>The parameter `encode` should similarly either only act as default if no `Content-Transfer-Encoding` is set yet (for each part separately!) or change the encoding if it is already present (especially since changing the charset the way it is done can also implicitly set the encoding to `base64`).
I think "encode" currently works only for the top level message
(i.e. not for attachments) - and it does something only if there
is not yet a "Content-Transfer-Encoding" header (if there is one,
it checks that conforms to the external "encode").

>Somehow I am confused about what the `charset` parameter actually does here. On the one hand, it seems to be used to decode the message if it is given as bytes. On the other hand, it is set as charset for all parts of the message that have no charset set yet.

As far as I can see, `MailHost` only operates on the top level message
(or the complete message) -- it does nothing for attachments.


>I guess having a multipart message with different charsets is not supported?

Not be the convenience parameters "charset/encode" -- those are for
the frequent simple cases.

Use other means to get multipart messages right.

> ...
>Also, not part of the changed lines in this PR, but `_string_transform` under Python 2 tries to encode bytes, which requires an implicit decoding and fails if there are non-ascii characters present.

`_string_transform` converts to a native string.

The motivation here is that "native string" is the "normal"
way for message creation. `_string_transform` thus is used to
transform the special case into the "normal" case - and it
expects, `charset` is appropriate for this transformation.

>What exactly is the effect of `encode`? If I understand correctly, it sets the `Content-Transfer-Encoding` of the message (*not* its parts in case of a multipart message) unless they are already set to something other than `7bit`.

What I understood: set `Content-Transfer-Encoding` to `encode`
(and perform the transformation) if the header is not yet set;
otherwise check the header value and `encode` are equal.


>However, at least together with passing a `charset` this still does not work. if I create a simple message like
>
>    To: test@test.de
>    From: test@test.de
>    Content-Type: text/plain
>
>    ü
>
>and call
>
>    host.send(msg, charset='utf-8', encode='8bit')
>
>this still results in the error `Message already encoded` because `_set_recursive_charset` changes the `Content-Transfer-Encoding` to `base64`.

A comment explains that this is by purpose: `encode` is an old
feature, partially superseeded by `charset`.
Modern `email` versions associate a "native" transfer encoding with `charset`.
If you "set" `charset`, `Message` automatically switches to
the corresponding transfer encoding. Thus, you should either
use `charset` or `encode` (but not both).

>
>> +        _writeBody = _handle_text
>+
>+    class FixedMessage(Message):
>+        def as_bytes(self, unixfrom=False, policy=None):
>+            policy = self.policy if policy is None else policy
>+            fp = BytesIO()
>+            g = FixedBytesGenerator(fp, mangle_from_=False, policy=policy)
>+            g.flatten(self, unixfrom=unixfrom)
>+            return fp.getvalue()
>+
>+    if hasattr(Compat32, 'message_factory'):
>+        fixed_policy = Compat32(message_factory=FixedMessage)
>+    else:
>+        fixed_policy = Compat32()  # 3.5 -- some tests will fail
>+
>+    message_from_string = partial(message_from_string, policy=fixed_policy)
>
>I guess these parts are actually meant to go into the PR for `zope.sendmail`?
>
>I do not think we need to patch `message_from_string` here. Instead, I would adopt the same strategy for `zope.sendmail` as `smtplib` does: Either pass a string that only contains ASCII or do the encoding yourself and pass bytes.

The `message_from_string` redefinition is necessary to
work around Python bug 41307. It ensures that created messages
use a "fixed `Message`" class -- one with a reliable `as_bytes`
method. Note that his PR depends on `as_bytes` because, it stipulates
that message at the level below `MailHost` are sequences of bytes.

>`zope.sendmail` should not be required to parse the message again (except for some header lines I guess, which should be ASCII).

You are right - it does this for convenience:
once, the message is parsed, one can easily access all headers.
Avoiding the parsing would require that we parse the headers
ourselves **BUT** MIME headers are complex.


>It would however be great if an error due to a string being passed that contains bytes with the eighth bit set were raised immediately and not in the second phase of the two-phase-commit.

Meanwhile, I follow your earlier proposal to encode via "utf-8" before
a `str` is passed as message body to `sendmail`.
While this can be wrong, it is what happens anyway when the message
gets queued. Rather than trying to be on the safe side in all cases
raising exceptions when this is impossible),
support the typical case conveniently. This may give wrong results
not detected via exceptions (but bad messages); in this case, the user
must use `bytes` rathen than `str` to take over control.

**Comment by viktordick** [comment]
@d-maurer Thanks for the explanations. Like I said, with this I am able to construct messages as required by setting the headers myself and not using the `encode` and `charset` parameters. I just think that it is not very clear what the parameters do and they seem to be inconsistent. Knowing that they are not meant to be used together and that `charset` is the newer option helps quite a bit. I guess this should go into the documentation.

> Those are for simple situations - a single message without "multipart". If you have complex messages do not use them. Do the necessary operations at the `email.message.Message` level (or even higher up).

But at least `charset` tries to apply itself to all parts of the message in case of a multipart message (see `_set_recursive_charset`).

> For Python 3 (where "bytes" is an explicit choice), I would say that a message provided as `bytes` should not be transformed at all: it already has the logical external format, We may just add/replace some information (e.g. the target addresses).

Agreed.

> `_string_transform` converts to a native string. The motivation here is that "native string" is the "normal" way for message creation. `_string_transform` thus is used to transform the special case into the "normal" case - and it expects, `charset` is appropriate for this transformation.

Sorry, I confused `six.text_type` with `six.string_type`. My bad.

> You are right - it does this for convenience: once, the message is parsed, one can easily access all headers. Avoiding the parsing would require that we parse the headers ourselves **BUT** MIME headers are complex.

I think it might be possible to split the message at the first double newline and pass only the header part to `message_from_bytes` or `message_from_string` in `zope.sendmail`. That way no encoding issues should appear and the patch to `zope.sendmail` will be minimal (if messages come in as bytes, keep them, otherwise convert them to bytes) and that library can concentrate on the transaction stuff. This makes the `FixedBytesGenerator` unnecessary at least in that library.

> Meanwhile, I follow your earlier proposal to encode via "utf-8" before a `str` is passed as message body to `sendmail`.

That was a desperate attempt because `Products.MailHost` was transforming my messages into strings even if I constructed them as bytes. But I guess it is still OK and will provide the least surprising behavior - as long as invalid UTF-8 is already caught in the first phase of the transaction. Failures in the second phase of a two-phase commit are nasty.


Anyhow, I guess we should first fix `zope.sendmail` to accept bytes (and encode using UTF-8 if strings are passed) before patching `Products.MailHost`? Part of the patch here actually belongs into a PR in `zope.sendmail`, does it not? I could give that a try.

**Comment by d-maurer** [comment]
Viktor Dick wrote at 2020-7-19 07:36 -0700:
>@d-maurer Thanks for the explanations. Like I said, with this I am able to construct messages as required by setting the headers myself and not using the `encode` and `charset` parameters. I just think that it is not very clear what the parameters do and they seem to be inconsistent. Knowing that they are not meant to be used together and that `charset` is the newer option helps quite a bit. I guess this should go into the documentation.

I do not mind if you work on "my" PR as well -- e.g. to improve
documentation.

> ...
>But at least `charset` tries to apply itself to all parts of the message in case of a multipart message (see `_set_recursive_charset`).

I may have overseen that.
I have only noticed the use of `charset` on the message as a whole
(to convert to native string) and the top level message.
But I have not really looked at multi-part messages.
>
> ...
>I think it might be possible to split the message at the first double newline and pass only the header part to `message_from_bytes` or `message_from_string` in `zope.sendmail`.

You are right.

>That way no encoding issues should appear and the patch to `zope.sendmail` will be minimal (if messages come in as bytes, keep them, otherwise convert them to bytes) and that library can concentrate on the transaction stuff. This makes the `FixedBytesGenerator` unnecessary at least in that library.

I did not see this consequwnce.

> ...
>Anyhow, I guess we should first fix `zope.sendmail` to accept bytes (and encode using UTF-8 if strings are passed) before patching `Products.MailHost`? Part of the patch here actually belongs into a PR in `zope.sendmail`, does it not? I could give that a try.

I agree.

File: src/Products/MailHost/tests/testMailHost.py [hunk 1]
```diff
@@ -692,6 +712,25 @@ def testStartQueueProcessorThread(self):
...
+""",
+                mto='user@example.com',
+                mfrom='zope@example.com',
+                subject='Hello world',
+                charset='utf-8')
+        transaction.commit()
+        self.assertTrue(mh.started_queue_processor_thread)
+        md = zope.sendmail.maildir.Maildir(self.smtp_queue_directory)
+        self.assertEqual(len(list(md)), 1)
+
```
Comment (jugmac00) [inline src/Products/MailHost/tests/testMailHost.py:733]: Just for my understanding. Why did you test `8bit` encoding in the `TestMailHostQueuing` test class and not in `TestMailHost` class?

**Comment by d-maurer** [comment]
Jürgen Gmach wrote at 2020-7-20 23:38 -0700:
>@jugmac00 commented on this pull request.
>
>> +    @unittest.skipUnless(PY36, 'requires Python 3.6+')
>+    def test_8bit_special(self):
>+        mh = self._makeOne('MailHost')
>+        mh.send(u"""\
>+Mime-Version: 1.0
>+Content-Type: text/plain; charset=UTF-8
>+Content-Transfer-Encoding: 8bit
>+
>+\u00e4
>+""",
>+                mto='user@example.com',
>+                mfrom='zope@example.com',
>+                subject='Hello world',
>+                charset='utf-8')
>+        transaction.commit()
>+        self.assertTrue(mh.started_queue_processor_thread)
>+        md = zope.sendmail.maildir.Maildir(self.smtp_queue_directory)
>+        self.assertEqual(len(list(md)), 1)
>+
>
>Just for my understanding. Why did you test `8bit` encoding in the `TestMailHostQueuing` test class and not in `TestMailHost` class?

The `TestMailHost` class has its own `8bit` encoding test (it was there
already, I did not add it).

I added the test above to ensure that my "overriding" of
`zope.sendmail` queueing does the right thing for the critical
`8bit` encoding case.
I had hoped that it would also show Viktor's problem -- but
this was impossible because the mailer used by the tests
is not `smtplib.SMTP.sendmail` (which would show the problem)
but a quite unfaithful emulation (especially, it does not
emulate `sendmail`s restriction that the message text must
be either bytes or ascii encodable).

File: src/Products/MailHost/MailHost.py [hunk 2]
```diff
@@ -385,9 +366,12 @@ def _string_transform(text, charset=None):
 
 
 def _mungeHeaders(messageText, mto=None, mfrom=None, subject=None,
-                  charset=None, msg_type=None):
+                  charset=None, msg_type=None, encode=None):
     """Sets missing message headers, and deletes Bcc.
-       returns fixed message, fixed mto and fixed mfrom"""
+       returns fixed message, fixed mto and fixed mfrom.
+
+       *messageText* can be either a ``Message`` or have
```
[the remark this answers is not in the document]
Comment (d-maurer) [reply src/Products/MailHost/MailHost.py:373]: Thank you! This part was lost indeed.

**Comment by d-maurer** [comment]
@dataflake From my point of view, this could get merged. But, I am unsure about the right way to state the dependency of `zope.sendmail >= 5.1`. I could modify `setup.py` (which now has `install_requires = [... 'zope.sendmail > 5`]`), but if I understood you correctly, you think that this may be too strong.

**Comment by dataflake** [comment]
@d-maurer I don't recall saying anything about changes to setup.py. The only thing about changing dependencies I can see here was the question about compatibility with Python 3.5, which must be retained because Zope 4.x still supports it.

Other than that changing the required version from `> 5` to `>= 5.1` is required here because there's code that relies on `zope.sendmail` 5.1

File: buildout.cfg [hunk 3]
```diff
@@ -1,11 +1,16 @@
 [buildout]
+extensions = mr.developer
```
Comment (dataflake) [inline buildout.cfg:2]: This needs to go now

File: buildout.cfg [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
     https://zopefoundation.github.io/Zope/releases/4.x/versions.cfg
 develop = .
 parts =
     test
     test_with_gs
     tox
+sources-dir = develop
+auto-checkout =
+
+[sources]
```
Comment (dataflake) [inline buildout.cfg:13]: This needs to go now

File: buildout.cfg (hunk 3 above)
Comment (d-maurer) [reply buildout.cfg]: Done

File: buildout.cfg (hunk 4 above)
Comment (d-maurer) [reply buildout.cfg]: Done

**Comment by dataflake** [comment]
Products.MailHost 4.10 is released

**Pull request merged**